#### Name: Sai Kumar Gandham
#### StudentID: IG45378

#### The twenty user groups dataset is know to have several words in it that make classification tasks easy, use NLTK's distribution function to find words that appear in 97.5% of of all documents or in less than 2.5% of the documents. Using a count vectorizer add those words to the stop_words parameter. What does this do to the accuracy of the classifier on the training set? On the testing set?

In [14]:
import nltk
from sklearn import datasets
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import numpy as np

# Downloading NLTK data
nltk.download('punkt')
nltk.download('stopwords')

# Fetching 20 Newsgroups dataset
twenty_users = datasets.fetch_20newsgroups()


[nltk_data] Downloading package punkt to
[nltk_data]     /Users/saikumargandham/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/saikumargandham/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [15]:
# Create CountVectorizer object
vectorizer = CountVectorizer()

# Fit and transform the corpus
X = vectorizer.fit_transform(twenty_users.data)

# Get the document frequency of each word
word_document_freq = np.sum(X > 0, axis=0).A1

# Calculate the threshold for document frequency
threshold_high = np.percentile(word_document_freq, 97.5)
threshold_low = np.percentile(word_document_freq, 2.5)

# Here We are getting words with document frequency greater than threshold_high or less than threshold_low
selected_words = [word for word, freq in zip(vectorizer.get_feature_names(), word_document_freq)
                  if freq > threshold_high or freq < threshold_low]

# Initializing stop words with the current stop words or an empty set if None
stop_words = vectorizer.get_stop_words() or set()

# Adding the selected words to stop words
stop_words.update(selected_words)


/Users/saikumargandham/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [16]:
# Setting stop words parameter of CountVectorizer
vectorizer.set_params(stop_words=stop_words)

# Fit and transform the corpus again with updated stop words
X_updated = vectorizer.fit_transform(twenty_users.data)

# Create a Multinomial Naive Bayes classifier
clf = MultinomialNB()

In [18]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_updated, twenty_users.target, test_size=0.2, random_state=42)

# Train the classifier on the training set
clf.fit(X_train, y_train)

# First Let's Predict on the training set
train_predictions = clf.predict(X_train)

# Calculating training accuracy
train_accuracy = accuracy_score(y_train, train_predictions)

# Predict on the testing set
test_predictions = clf.predict(X_test)

# Calculate testing accuracy
test_accuracy = accuracy_score(y_test, test_predictions)

# Print accuracies
print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)


Training Accuracy: 0.9804441498176997
Testing Accuracy: 0.866548828988069


High Training Accuracy: The model learned the training data really well (98% accuracy).

Slightly Lower Testing Accuracy: When tested on new data, the accuracy dropped a bit (87%). It's still good, but not as high as the training accuracy.

Potential Overfitting: The big difference between how well the model did on the training and testing data could mean it's too focused on the training data. This might make it struggle when dealing with new, unseen information. We might need to look more closely at the model and make some changes to help it perform better with new data.



